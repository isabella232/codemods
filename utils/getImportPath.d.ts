export declare function getImportPath(schemaPathFromArgs?: string, cwd?: string): Promise<string>;
