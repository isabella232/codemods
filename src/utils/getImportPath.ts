import { getGenerators } from '@prisma/sdk'
import { getSchemaPathInternal } from '@prisma/sdk/dist/cli/getSchema';
import path from 'path';

export async function getImportPath(schemaPathFromArgs?: string, cwd?: string){
  if(!cwd){
    return '@prisma/client'
  }
  const schemaPath = await getSchemaPathInternal(schemaPathFromArgs, {cwd})
  // console.log({schemaPath});
  if(schemaPath){
    const generators = await getGenerators({schemaPath})
    const generator = generators[0]
    if(generator){
      const isCustom = generator.options?.generator?.isCustomOutput
      const customImportPath = path.relative(cwd, generator.options?.generator.output!)
      generator.stop()
      return isCustom ? customImportPath : '@prisma/client'
    }
  }
  return '@prisma/client'
}

function prefixRelativePathIfNecessary(relativePath: string): string {
  if (relativePath.startsWith('..')) {
    return relativePath
  }

  return `./${relativePath}`
}
