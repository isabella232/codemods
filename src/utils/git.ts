'use strict'

import fs from 'fs'
import path from 'path'
import { execSync, ExecSyncOptions} from 'child_process'

// Prevent from failing on windows
const nullPath = /^win/.test(process.platform) ? 'nul' : '/dev/null'

// Consider EOL as \n because either Windows or *nix, this escape char will be there
const EOL = /\r?\n/

export function isGit(dir: string) {
  return fs.existsSync(path.join(dir, '.git'))
}

export function untracked (repo: string, opts?: ExecSyncOptions) {
  return status(repo, opts).untracked
}

export function dirty (repo: string, opts?: ExecSyncOptions) {
  return status(repo, opts).dirty
}

export function branch (repo: string, opts?: ExecSyncOptions) {
  opts = opts || {}
  try {
    const stdout = execSync('git show-ref >' + nullPath + ' 2>&1 && git rev-parse --abbrev-ref HEAD', {cwd: repo, maxBuffer: opts.maxBuffer}).toString()
    return stdout.trim()
  } catch (err) {
    if (err.code === 'ENOBUFS') throw err
    return null // most likely the git repo doesn't have any commits yet
  }
}

export function ahead (repo: string, opts?: ExecSyncOptions) {
  opts = opts || {}
  try {
    let stdout = execSync('git show-ref >' + nullPath + ' 2>&1 && git rev-list HEAD --not --remotes', {cwd: repo, maxBuffer: opts.maxBuffer}).toString()
    stdout = stdout.trim()
    return !stdout ? 0 : stdout.split(EOL).length
  } catch (err) {
    if (err.code === 'ENOBUFS') throw err
    return NaN
  }
}

// Throws error
export function status(repo: string, opts?: ExecSyncOptions) {
  opts = opts || {}
  const stdout = execSync('git status -s', {cwd: repo, maxBuffer: opts.maxBuffer}).toString()
  const status = { dirty: 0, untracked: 0 }
  stdout.trim().split(EOL).filter(Boolean).forEach(function (file) {
    if (file.substr(0, 2) === '??') status.untracked++
    else status.dirty++
  })
  return status
}

// Throws error
export function commit (repo: string, opts?: ExecSyncOptions) {
  opts = opts || {}
  const stdout = execSync('git rev-parse --short HEAD', {cwd: repo, maxBuffer: opts.maxBuffer}).toString()
  const commitHash = stdout.trim()
  return commitHash
}

// Throws error
export function stashes (repo: string, opts?: ExecSyncOptions) {
  opts = opts || {}
  const stdout = execSync('git stash list', {cwd: repo, maxBuffer: opts.maxBuffer}).toString()
  const stashes = stdout.trim().split(EOL).filter(Boolean)
  return stashes.length
}

// Throws error
export function message (repo: string, opts?: ExecSyncOptions) {
  opts = opts || {}
  return execSync('git log -1 --pretty=%B', {cwd: repo, maxBuffer: opts.maxBuffer}).toString().trim()
}

export function check(repo: string, opts?: ExecSyncOptions) {
  const statusResult = status(repo, opts)
  return {
    branch: branch(repo, opts),
    ahead: ahead(repo, opts),
    dirty: statusResult.dirty,
    untracked: statusResult.untracked,
    stashes: stashes(repo, opts)
  }
}
