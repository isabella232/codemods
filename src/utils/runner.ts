import execa from "execa";
import path from "path";
export function run(transform: string, paths: string, options: {dry: boolean, debug: boolean}){
  return execa(
    path.join(__dirname, "..", "..", "node_modules", ".bin", "jscodeshift"),
    [
      options.dry ? "--dry" : "",
      options.debug ? "--verbose=2" : "",
      "--print",
      "--run-in-band",
      "-t",
      transform,
      "--extensions=ts",
      "--parser=ts",
      paths,
    ],
    {
      encoding: "utf8",
    }
  );
}
export function buildRunner(transform: string) {
  return (paths: string, options: {dry: boolean, debug: boolean}) => run(transform, paths, options)
}

