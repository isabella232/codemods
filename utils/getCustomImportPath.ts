import { getConfig, getGenerators } from '@prisma/sdk'
import { getSchemaPathInternal } from '@prisma/sdk/dist/cli/getSchema';
import path from 'path';
import fs from 'fs'

interface Options {
  schemaPathFromArgs?: string, 
  cwd?: string
}
export async function getCustomImportPath(options?: Options){
  if(!options?.cwd){
    return null
  }
  const schemaPath = await getSchemaPathInternal(options?.schemaPathFromArgs, {cwd: options.cwd})
  const datamodel = schemaPath && fs.readFileSync(schemaPath, {encoding: 'utf8'})
  if(schemaPath){
    const config = await getConfig({datamodel, ignoreEnvVarErrors: true})
    const generator = config.generators[0]

    if(generator){
      const isCustom = generator?.output
      const customImportPath = isCustom && path.relative(options.cwd, generator.output)
      return isCustom ? customImportPath.replace(/\.\.\//g, '') : null
    }
  }
  return null
}
