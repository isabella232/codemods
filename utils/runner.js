"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildRunner = exports.run = void 0;
const execa_1 = __importDefault(require("execa"));
function run(transform, paths, options) {
    const args = [
        options.dry ? "--dry" : "",
        options.debug ? "--verbose=2" : "",
        options.print ? "--print" : "",
        options.runInBand ? "--run-in-band" : "",
        `--ignore-pattern="node_modules/`,
        "-t",
        transform,
        "--extensions=ts",
        "--parser=ts",
        paths,
    ];
    const jscodeshiftExecutable = require.resolve('.bin/jscodeshift');
    // console.log(args);
    return execa_1.default(jscodeshiftExecutable, args, {
        encoding: "utf8",
    });
}
exports.run = run;
function buildRunner(transform) {
    return (paths, options) => run(transform, paths, options);
}
exports.buildRunner = buildRunner;
//# sourceMappingURL=runner.js.map