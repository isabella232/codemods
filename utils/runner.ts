import execa from "execa";
import fs from 'fs';
import path from "path";
export interface Options {
  dry: boolean;
  debug: boolean;
  ignorePattern?: string;
  print?: boolean;
  runInBand?: boolean;
}
export const transformerDirectory = path.join(__dirname, "../", "transforms");
export const jscodeshiftExecutable = require.resolve(".bin/jscodeshift");

export function runTransform({
  projectDir,
  flags,
  transformer,
  customImportPath,
  testMode
}: {
  projectDir: string;
  transformer: string;
  customImportPath?: string;
  flags: { dry?: boolean; print?: boolean, runInBand?: boolean };
  testMode?: boolean
}) {
  const transformerPath = path.join(transformerDirectory, `${transformer}.js`);

  let args = [];

  const { dry, print, runInBand } = flags;

  if (dry) {
    args.push("--dry");
  }
  if (print) {
    args.push("--print");
  }
  if (runInBand) {
    args.push("--run-in-band");
  }
  if(testMode){
    args.push("--verbose=2");
  } else {
    args.push("--verbose=0");
  }
  args.push("--no-babel")
  if(customImportPath){
    args.push(`--ignore-pattern=**/${customImportPath}/**`);
  }
  if(typeof projectDir === 'string'){
    if(fs.existsSync(path.join(projectDir, '.gitignore'))){
      args.push(`--ignore-config=${path.join(projectDir, '.gitignore')}`)
    }
    else if(fs.existsSync(path.join(projectDir,'..', '.gitignore'))){
      args.push(`--ignore-config=${path.join(projectDir, '..','.gitignore')}`)
    }
  }
  args.push("--ignore-pattern=**/node_modules/**");
  // TODO Check TSX parser
  args.push("--extensions=tsx,ts,js,tsx");
  args.push("--parser=tsx");

  args = args.concat(["--transform", transformerPath]);

  args = args.concat(projectDir);

  console.log(`Executing command: jscodeshift ${args.join(" ")}`);

  const result = execa.sync(jscodeshiftExecutable, args, {
    stdio: testMode ? "pipe" : "inherit",
    stripFinalNewline: false,
  });

  if (result.stderr) {
    throw result.stderr;
  }
  return result
}
