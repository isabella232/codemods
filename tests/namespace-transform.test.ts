import path from "path";
import { buildRunner } from "../src/utils/runner";

const transform = require.resolve("../lib/transforms/namespace");
const testRunner = buildRunner(transform)
const options = {dry: true, debug: false}
test("minimal example", async () => {
  const fixturePath = path.join(__dirname, "./__fixtures__/namespace.input.ts")
  const result = await testRunner(fixturePath, options)
  expect(result.stdout).toMatchSnapshot();
});

test("example with sql", async () => {
  const fixturePath = path.join(__dirname, "./__fixtures__/sql.input.ts")
  const result = await testRunner(fixturePath, options)
  expect(result.stdout).toMatchSnapshot();
});

test("example with error classes", async () => {
  const fixturePath = path.join(__dirname, "./__fixtures__/errorImports.input.ts")
  const result = await testRunner(fixturePath, options)
  expect(result.stdout).toMatchSnapshot();
});

test("example with decimal", async () => {
  const fixturePath = path.join(__dirname, "./__fixtures__/decimal.input.ts")
  const result = await testRunner(fixturePath, options)
  expect(result.stdout).toMatchSnapshot();
});

test("js example with decimal", async () => {
  const fixturePath = path.join(__dirname, "./__fixtures__/decimal.input.js")
  const result = await testRunner(fixturePath, options)
  expect(result.stdout).toMatchSnapshot();
});
