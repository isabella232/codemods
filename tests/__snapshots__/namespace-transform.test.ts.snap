// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inputs decimal.js 1`] = `
import { PrismaClient, Prisma } from '@prisma/client';

async function main() {
  const prisma = new PrismaClient()

  const a = await prisma.a.findFirst()
  const b = await prisma.b.findFirst({
    where: {
      decFloat: new Prisma.Decimal('1.23')
    }
  })
  const c = await prisma.c.findFirst()
  const d = await prisma.d.findFirst()
  const e = await prisma.e.findFirst()

}

main()

`;

exports[`inputs decimal.ts 1`] = `
import { PrismaClient, Prisma } from '@prisma/client';

async function main() {
  const prisma = new PrismaClient()

  const a: null | {
    id: string
    email: string
    name: string | null
    int: number
    sInt: number
    bInt: BigInt
    serial: number
    sSerial: number
    bSerial: number
    inc_int: number
    inc_sInt: number
    inc_bInt: BigInt
  } = await prisma.a.findFirst()
  const b = await prisma.b.findFirst({
    where: {
      decFloat: new Prisma.Decimal('1.23')
    }
  })
  const c = await prisma.c.findFirst()
  const d: null | {
    id: string
    bool: boolean
    byteA: Buffer
    xml: string
    json: Prisma.JsonValue
    jsonb: Prisma.JsonValue
  } = await prisma.d.findFirst()
  const e = await prisma.e.findFirst()

}

main()

`;

exports[`inputs errorImports.ts 1`] = `
const {
  Prisma,
  PrismaClient
} = require('./node_modules/@prisma/client')

test('blog', async () => {
  const requests: any[] = []
  const db = new PrismaClient({
    errorFormat: 'colorless',
    __internal: {
      hooks: {
        beforeRequest: (request) => requests.push(request),
      },
    },
  })

  if (!Prisma.prismaVersion || !Prisma.prismaVersion.client) {
    throw new Error(\`prismaVersion missing: \${JSON.stringify(Prisma.prismaVersion)}\`)
  }

  // Test connecting and disconnecting all the time
  await db.user.findMany()
  const posts = await db.user
    .findOne({
      where: {
        email: 'a@a.de',
      },
    })
    .posts()

  expect(posts.length).toBe(0)
  db.$disconnect()
  expect(requests.length).toBe(2)

  await db.user.findMany()
  db.$disconnect()
  expect(requests.length).toBe(3)

  const count = await db.user.count()
  expect(typeof count === 'number').toBe(true)

  const paramCount = await db.user.count({
    take: 10000,
  })

  expect(typeof paramCount === 'number').toBe(true)

  db.$connect()
  await db.$disconnect()

  await new Promise((r) => setTimeout(r, 200))
  db.$connect()

  const userPromise = db.user.findMany()
  await userPromise
  // @ts-ignore

  await db.$disconnect()

  await db.$connect()

  /**
   * queryRaw
   */

  // Test queryRaw(string)
  const rawQuery = await db.$queryRaw('SELECT 1')
  expect(rawQuery[0]['1']).toBe(1)

  // Test queryRaw(string, values)
  const rawQueryWithValues = await db.$queryRaw(
    'SELECT $1 AS name, $2 AS id',
    'Alice',
    42,
  )

  expect(rawQueryWithValues[0]).toEqual({
    name: 'Alice',
    id: 42,
  })

  // Test queryRaw\`\`
  const rawQueryTemplate = await db.$queryRaw\`SELECT 1\`
  expect(rawQueryTemplate[0]['1']).toBe(1)

  // Test queryRaw\`\` with \${param}
  const rawQueryTemplateWithParams = await db.$queryRaw\`SELECT * FROM User WHERE name = \${'Alice'}\`
  expect(rawQueryTemplateWithParams[0].name).toBe('Alice')

  // Test queryRaw\`\` with prisma.sql\`\`
  const rawQueryTemplateFromSqlTemplate = await db.$queryRaw(
    Prisma.sql\`
      SELECT \${Prisma.join([Prisma.raw('email'), Prisma.raw('id'), Prisma.raw('name')])}
      FROM \${Prisma.raw('User')}
      \${Prisma.sql\`WHERE name = \${'Alice'}\`}
      \${Prisma.empty}
    \`,
  )
  expect(rawQueryTemplateFromSqlTemplate[0].name).toBe('Alice')

  /**
   * .$executeRaw(
   */

  // Test .$executeRaw((string)
  const executeRaw = await db.$executeRaw(
    'UPDATE User SET name = $1 WHERE id = $2',
    'name',
    'id',
  )
  expect(executeRaw).toBe(0)

  // Test .$executeRaw((string, values)
  const executeRawWithValues = await db.$executeRaw(
    'UPDATE User SET name = $1 WHERE id = $2',
    'Alice',
    'id',
  )
  expect(executeRawWithValues).toBe(0)

  // Test $executeRaw
  const $executeRawTemplate = await db.$executeRaw\`UPDATE User SET name = \${'name'} WHERE id = \${'id'}\`
  expect($executeRawTemplate).toBe(0)

  // Test validation errors
  let validationError
  try {
    await db.post.create({
      data: {},
    })
  } catch (e) {
    validationError = e
  } finally {
    if (
      !validationError ||
      !(validationError instanceof Prisma.PrismaClientValidationError)
    ) {
      throw new Error(\`Validation error is incorrect\`)
    }
  }

  // Test known request error
  let knownRequestError
  try {
    const result = await db.user.create({
      data: {
        email: 'a@a.de',
        name: 'Alice',
      },
    })
  } catch (e) {
    knownRequestError = e
  } finally {
    if (
      !knownRequestError ||
      !(knownRequestError instanceof Prisma.PrismaClientKnownRequestError)
    ) {
      throw new Error(\`Known request error is incorrect\`)
    } else {
      if (!knownRequestError.message.includes('.user.create()')) {
        throw new Error(\`Invalid error: \${knownRequestError.message}\`)
      }
    }
  }

  // relation query where not null
  const relationWhereNotNull = await db.user.findMany({
    where: {
      profile: {
        bio: { not: null },
      },
    },
  })
  expect(relationWhereNotNull).toEqual([])

  db.$disconnect()
})

`;

exports[`inputs keyof.ts 1`] = `
import { PrismaClient, Prisma } from '@prisma/client';
type X = Required<Prisma.UserArgs>
type Y = {
  hello: Prisma.UserArgs
}
type Z = {
  [T in keyof Prisma.UserArgs]: Prisma.UserArgs[T]
}[keyof Prisma.UserArgs]

function a(a: Prisma.UserArgs): Prisma.UserArgs {
  return null as any
}
function b<T extends Prisma.UserArgs>() {
}
async function main() {
  const prisma = new PrismaClient()
  await prisma.user.update({
    where: {
    },
    data: {
      posts: {
      }
    }
  })
  prisma.$disconnect()
}
main()
`;

exports[`inputs minimal.ts 1`] = `
import { Prisma, PrismaClient } from '@prisma/client'
import * as fs from 'fs'
function test(){
    const why = fs
}
// in the code
const args: Prisma.UserArgs
`;

exports[`inputs sql.ts 1`] = `
import { PrismaClient, Prisma } from '@prisma/client';

// tslint:disable

// This file will not be executed, just compiled to check if the typings are valid
async function main() {
  const prisma = new PrismaClient({
    log: [
      {
        emit: 'event',
        level: 'query',
      },
    ],
    datasources: {
      db: {
        url: 'file:dev.db',
      },
    },
  })

  prisma.on('query', (a) => {
    //
  })

  Prisma.prismaVersion.client

  const x: Prisma.Sql = Prisma.sql\`SELECT * FROM \${Prisma.raw('User')} WHERE 'id' in \${Prisma.join([
    1,
    2,
    3,
  ])} \${Prisma.empty} \`

  const queryRaw1 = await prisma.$queryRaw\`SELECT * FROM User WHERE id = 1\`
  const queryRaw2 = await prisma.$queryRaw\`SELECT * FROM User WHERE id = \${1}\`
  const queryRaw3 = await prisma.$queryRaw(
    \`SELECT * FROM User WHERE id = $1\`,
    2,
  )
  const queryRaw4 = await prisma.$queryRaw(
    Prisma.sql\`SELECT * FROM User WHERE id = \${1}\`,
  )
  const queryRaw5 = await prisma.$queryRaw(
    Prisma.sql\`SELECT * FROM User \${Prisma.sql\`WHERE id = \${1}\`}\`,
  )

  const executeRaw1 = await prisma.$executeRaw\`SELECT * FROM User WHERE id = 1\`
  const executeRaw2 = await prisma.$executeRaw\`SELECT * FROM User WHERE id = \${1}\`
  const executeRaw3 = await prisma.$executeRaw(
    \`SELECT * FROM User WHERE id = $1\`,
    2,
  )
  const executeRaw4 = await prisma.$executeRaw(
    Prisma.sql\`SELECT * FROM User WHERE id = \${1}\`,
  )
  const executeRaw5 = await prisma.$executeRaw(
    Prisma.sql\`SELECT * FROM User \${Prisma.sql\`WHERE id = \${1}\`}\`,
  )

  const result1 = await prisma.user.findMany({
    where: {
      posts: {
        some: {
          author: {
            AND: {
              id: '5',
              posts: {
                some: {
                  author: {
                    posts: {
                      some: {
                        title: '5',
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  })

  result1[0]!.email
  result1[0]!.id
  result1[0]!.name

  const result2: {
    id: string
    createdAt: Date
    updatedAt: Date
    published: boolean
    title: string
    content: string | null
    author: Prisma.User | null
  } | null = await prisma.post.findOne({
    where: {
      id: '',
    },
    include: {
      author: true,
    },
  })

  const result3: 'Please either choose \`select\` or \`include\`' = await prisma.post.findMany(
    {
      select: {},
      include: {},
    },
  )

  const result4: Array<{
    id: string
    author: {
      name: string | null
    } | null
  }> = await prisma.post.findMany({
    select: {
      id: true,
      author: {
        select: {
          name: true,
        },
      },
    },
  })

  const result5: Prisma.Post = await prisma.post.create({
    data: {
      published: false,
      title: 'Title',
    },
  })

  await prisma.post.delete({
    where: {
      id: '',
    },
  })

  await prisma.post.upsert({
    create: {
      published: false,
      title: 'Title',
    },
    update: {
      published: true,
    },
    where: {
      id: '6',
    },
  })

  await prisma.post.updateMany({
    data: {
      published: false,
    },
  })

  const count: number = await prisma.post.count({
    where: {
      published: false,
    },
  })

  const $disconnect: Promise<void> = prisma.$disconnect()

  // Trick to define a "positive" test, if "include" is NOT in "FindManyMachineDataArgs"
  type X = keyof Prisma.FindManyMachineDataArgs
  type Y = 'include' extends X ? number : string
  const y: Y = 'string'

  // Test for https://github.com/prisma/prisma-client-js/issues/615
  const users = await prisma.user.findMany({
    include: {
      posts: {
        include: {
          author: true,
        },
        orderBy: {
          title: 'asc',
        },
      },
    },
  })

  const id = users[0].posts[0].author?.id

  const like = await prisma.like.findOne({
    where: {
      userId_postId: {
        postId: '',
        userId: '',
      },
    },
    include: { post: true },
  })

  like!.post

  const like2 = await prisma.like.upsert({
    where: {
      userId_postId: {
        userId: '',
        postId: '',
      },
    },
    create: {
      post: { connect: { id: '' } },
      user: { connect: { id: '' } },
    },
    update: {},
    include: { post: true },
  })

  like2!.post

  // make sure, that null is not allowed for this type
  type LikeUpdateIdType = Prisma.LikeUpdateManyArgs['data']['id']
  type AllowsNull = null extends LikeUpdateIdType ? true : false
  const allowsNull: AllowsNull = false

  // check if listing of \`set\` is done in nested relations
  // https://github.com/prisma/prisma/issues/3497
  await prisma.user.update({
    where: {
      id: '6',
    },
    data: {
      posts: {
        update: {
          data: {
            title: 'something',
          },
          where: {
            id: 'whatever',
          },
        },
      },
    },
  })

  await prisma.user.update({
    where: {
      id: '6',
    },
    data: {
      posts: {
        updateMany: {
          data: {
            title: 'something',
          },
          where: {
            id: 'whatever',
          },
        },
      },
    },
  })

  await prisma.post.update({
    where: {
      id: '6',
    },
    data: {
      author: {
        update: {
          name: 'something',
        },
      },
    },
  })
}

main().catch((e) => {
  console.error(e)
})

`;

exports[`projects minimal 1`] = `
import { PrismaClient, Prisma } from './generated/client';

// tslint:disable

// This file will not be executed, just compiled to check if the typings are valid
async function main() {
  const prisma = new PrismaClient({
    log: [
      {
        emit: 'event',
        level: 'query',
      },
    ],
    datasources: {
      db: {
        url: 'file:dev.db',
      },
    },
  })

  prisma.on('query', (a) => {
    //
  })

  Prisma.prismaVersion.client

  const x: Prisma.Sql = Prisma.sql\`SELECT * FROM \${Prisma.raw('User')} WHERE 'id' in \${Prisma.join([
    1,
    2,
    3,
  ])} \${Prisma.empty} \`

  const queryRaw1 = await prisma.$queryRaw\`SELECT * FROM User WHERE id = 1\`
  const queryRaw2 = await prisma.$queryRaw\`SELECT * FROM User WHERE id = \${1}\`
  const queryRaw3 = await prisma.$queryRaw(
    \`SELECT * FROM User WHERE id = $1\`,
    2,
  )
  const queryRaw4 = await prisma.$queryRaw(
    Prisma.sql\`SELECT * FROM User WHERE id = \${1}\`,
  )
  const queryRaw5 = await prisma.$queryRaw(
    Prisma.sql\`SELECT * FROM User \${Prisma.sql\`WHERE id = \${1}\`}\`,
  )

  const executeRaw1 = await prisma.$executeRaw\`SELECT * FROM User WHERE id = 1\`
  const executeRaw2 = await prisma.$executeRaw\`SELECT * FROM User WHERE id = \${1}\`
  const executeRaw3 = await prisma.$executeRaw(
    \`SELECT * FROM User WHERE id = $1\`,
    2,
  )
  const executeRaw4 = await prisma.$executeRaw(
    Prisma.sql\`SELECT * FROM User WHERE id = \${1}\`,
  )
  const executeRaw5 = await prisma.$executeRaw(
    Prisma.sql\`SELECT * FROM User \${Prisma.sql\`WHERE id = \${1}\`}\`,
  )

  const result1 = await prisma.user.findMany({
    where: {
      posts: {
        some: {
          author: {
            AND: {
              id: '5',
              posts: {
                some: {
                  author: {
                    posts: {
                      some: {
                        title: '5',
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  })

  result1[0]!.email
  result1[0]!.id
  result1[0]!.name

  const result2: {
    id: string
    createdAt: Date
    updatedAt: Date
    published: boolean
    title: string
    content: string | null
    author: Prisma.User | null
  } | null = await prisma.post.findOne({
    where: {
      id: '',
    },
    include: {
      author: true,
    },
  })

  const result3: 'Please either choose \`select\` or \`include\`' = await prisma.post.findMany(
    {
      select: {},
      include: {},
    },
  )

  const result4: Array<{
    id: string
    author: {
      name: string | null
    } | null
  }> = await prisma.post.findMany({
    select: {
      id: true,
      author: {
        select: {
          name: true,
        },
      },
    },
  })

  const result5: Prisma.Post = await prisma.post.create({
    data: {
      published: false,
      title: 'Title',
    },
  })

  await prisma.post.delete({
    where: {
      id: '',
    },
  })

  await prisma.post.upsert({
    create: {
      published: false,
      title: 'Title',
    },
    update: {
      published: true,
    },
    where: {
      id: '6',
    },
  })

  await prisma.post.updateMany({
    data: {
      published: false,
    },
  })

  const count: number = await prisma.post.count({
    where: {
      published: false,
    },
  })

  const $disconnect: Promise<void> = prisma.$disconnect()

  // Trick to define a "positive" test, if "include" is NOT in "FindManyMachineDataArgs"
  type X = keyof Prisma.FindManyMachineDataArgs
  type Y = 'include' extends X ? number : string
  const y: Y = 'string'

  // Test for https://github.com/prisma/prisma-client-js/issues/615
  const users = await prisma.user.findMany({
    include: {
      posts: {
        include: {
          author: true,
        },
        orderBy: {
          title: 'asc',
        },
      },
    },
  })

  const id = users[0].posts[0].author?.id

  const like = await prisma.like.findOne({
    where: {
      userId_postId: {
        postId: '',
        userId: '',
      },
    },
    include: { post: true },
  })

  like!.post

  const like2 = await prisma.like.upsert({
    where: {
      userId_postId: {
        userId: '',
        postId: '',
      },
    },
    create: {
      post: { connect: { id: '' } },
      user: { connect: { id: '' } },
    },
    update: {},
    include: { post: true },
  })

  like2!.post

  // make sure, that null is not allowed for this type
  type LikeUpdateIdType = Prisma.LikeUpdateManyArgs['data']['id']
  type AllowsNull = null extends LikeUpdateIdType ? true : false
  const allowsNull: AllowsNull = false

  // check if listing of \`set\` is done in nested relations
  // https://github.com/prisma/prisma/issues/3497
  await prisma.user.update({
    where: {
      id: '6',
    },
    data: {
      posts: {
        update: {
          data: {
            title: 'something',
          },
          where: {
            id: 'whatever',
          },
        },
      },
    },
  })

  await prisma.user.update({
    where: {
      id: '6',
    },
    data: {
      posts: {
        updateMany: {
          data: {
            title: 'something',
          },
          where: {
            id: 'whatever',
          },
        },
      },
    },
  })

  await prisma.post.update({
    where: {
      id: '6',
    },
    data: {
      author: {
        update: {
          name: 'something',
        },
      },
    },
  })
}

main().catch((e) => {
  console.error(e)
})

import { PrismaClient, Prisma } from '../generated/client';

// tslint:disable

// This file will not be executed, just compiled to check if the typings are valid
async function main() {
  const prisma = new PrismaClient({
    log: [
      {
        emit: 'event',
        level: 'query',
      },
    ],
    datasources: {
      db: {
        url: 'file:dev.db',
      },
    },
  })

  prisma.on('query', (a) => {
    //
  })

  Prisma.prismaVersion.client

  const x: Prisma.Sql = Prisma.sql\`SELECT * FROM \${Prisma.raw('User')} WHERE 'id' in \${Prisma.join([
    1,
    2,
    3,
  ])} \${Prisma.empty} \`

  const queryRaw1 = await prisma.$queryRaw\`SELECT * FROM User WHERE id = 1\`
  const queryRaw2 = await prisma.$queryRaw\`SELECT * FROM User WHERE id = \${1}\`
  const queryRaw3 = await prisma.$queryRaw(
    \`SELECT * FROM User WHERE id = $1\`,
    2,
  )
  const queryRaw4 = await prisma.$queryRaw(
    Prisma.sql\`SELECT * FROM User WHERE id = \${1}\`,
  )
  const queryRaw5 = await prisma.$queryRaw(
    Prisma.sql\`SELECT * FROM User \${Prisma.sql\`WHERE id = \${1}\`}\`,
  )

  const executeRaw1 = await prisma.$executeRaw\`SELECT * FROM User WHERE id = 1\`
  const executeRaw2 = await prisma.$executeRaw\`SELECT * FROM User WHERE id = \${1}\`
  const executeRaw3 = await prisma.$executeRaw(
    \`SELECT * FROM User WHERE id = $1\`,
    2,
  )
  const executeRaw4 = await prisma.$executeRaw(
    Prisma.sql\`SELECT * FROM User WHERE id = \${1}\`,
  )
  const executeRaw5 = await prisma.$executeRaw(
    Prisma.sql\`SELECT * FROM User \${Prisma.sql\`WHERE id = \${1}\`}\`,
  )

  const result1 = await prisma.user.findMany({
    where: {
      posts: {
        some: {
          author: {
            AND: {
              id: '5',
              posts: {
                some: {
                  author: {
                    posts: {
                      some: {
                        title: '5',
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  })

  result1[0]!.email
  result1[0]!.id
  result1[0]!.name

  const result2: {
    id: string
    createdAt: Date
    updatedAt: Date
    published: boolean
    title: string
    content: string | null
    author: Prisma.User | null
  } | null = await prisma.post.findOne({
    where: {
      id: '',
    },
    include: {
      author: true,
    },
  })

  const result3: 'Please either choose \`select\` or \`include\`' = await prisma.post.findMany(
    {
      select: {},
      include: {},
    },
  )

  const result4: Array<{
    id: string
    author: {
      name: string | null
    } | null
  }> = await prisma.post.findMany({
    select: {
      id: true,
      author: {
        select: {
          name: true,
        },
      },
    },
  })

  const result5: Prisma.Post = await prisma.post.create({
    data: {
      published: false,
      title: 'Title',
    },
  })

  await prisma.post.delete({
    where: {
      id: '',
    },
  })

  await prisma.post.upsert({
    create: {
      published: false,
      title: 'Title',
    },
    update: {
      published: true,
    },
    where: {
      id: '6',
    },
  })

  await prisma.post.updateMany({
    data: {
      published: false,
    },
  })

  const count: number = await prisma.post.count({
    where: {
      published: false,
    },
  })

  const $disconnect: Promise<void> = prisma.$disconnect()

  // Trick to define a "positive" test, if "include" is NOT in "FindManyMachineDataArgs"
  type X = keyof Prisma.FindManyMachineDataArgs
  type Y = 'include' extends X ? number : string
  const y: Y = 'string'

  // Test for https://github.com/prisma/prisma-client-js/issues/615
  const users = await prisma.user.findMany({
    include: {
      posts: {
        include: {
          author: true,
        },
        orderBy: {
          title: 'asc',
        },
      },
    },
  })

  const id = users[0].posts[0].author?.id

  const like = await prisma.like.findOne({
    where: {
      userId_postId: {
        postId: '',
        userId: '',
      },
    },
    include: { post: true },
  })

  like!.post

  const like2 = await prisma.like.upsert({
    where: {
      userId_postId: {
        userId: '',
        postId: '',
      },
    },
    create: {
      post: { connect: { id: '' } },
      user: { connect: { id: '' } },
    },
    update: {},
    include: { post: true },
  })

  like2!.post

  // make sure, that null is not allowed for this type
  type LikeUpdateIdType = Prisma.LikeUpdateManyArgs['data']['id']
  type AllowsNull = null extends LikeUpdateIdType ? true : false
  const allowsNull: AllowsNull = false

  // check if listing of \`set\` is done in nested relations
  // https://github.com/prisma/prisma/issues/3497
  await prisma.user.update({
    where: {
      id: '6',
    },
    data: {
      posts: {
        update: {
          data: {
            title: 'something',
          },
          where: {
            id: 'whatever',
          },
        },
      },
    },
  })

  await prisma.user.update({
    where: {
      id: '6',
    },
    data: {
      posts: {
        updateMany: {
          data: {
            title: 'something',
          },
          where: {
            id: 'whatever',
          },
        },
      },
    },
  })

  await prisma.post.update({
    where: {
      id: '6',
    },
    data: {
      author: {
        update: {
          name: 'something',
        },
      },
    },
  })
}

main().catch((e) => {
  console.error(e)
})

`;
